fix: krm exec function working dir

aabouzaid opened the pull request:
Based on @seh [tip](https://github.com/kubernetes-sigs/kustomize/issues/4347#issuecomment-1140290946) in #4347

This change makes a deep copy from the loader config instead of passing a reference to the config.

How to test: Follow the steps in #4347.

I've added 2 tests:
- **TestFnExecTransformer:** This test will pass in all cases because before the fix the base path is used and there is only the base and no overlay.
- **TestFnExecTransformerWithOverlay:**
  - **Before the fix:** [The test failed](https://github.com/aabouzaid/kustomize/runs/7313566978?check_suite_focus=true#step:4:5295) with exact error as in 4347 because the working dir is set to the base path.
  - **After the fix:** [The test passed](https://github.com/aabouzaid/kustomize/runs/7313727469?check_suite_focus=true#step:4:5295) where the working dir is updated with the overlay path.

aabouzaid commented:
/check-cla

seh commented:
@KnVerey, does the implementation match your expectations?

@aabouzaid, have you found an existing test that we can augment or duplicate to test this change?

natasha41575 commented on api/internal/target/kusttarget.go:
```diff
@@ -45,6 +45,7 @@ func NewKustTarget(
 	pLdr *loader.Loader) *KustTarget {
 	pLdrCopy := *pLdr
+	pLdrCopy.SetConfig()
 	pLdrCopy.SetWorkDir(ldr.Root())
```
I think it might be better to create a `Loader.DeepCopy` method that combines lines 47-49 (and fold your `Loader.SetConfig` into it). WDYT?

aabouzaid replied:
Good idea, I've implemented it :+1:

natasha41575 reviewed (commented):
Please feel free to add a unit test, this approach seems fine to me. Thanks!

natasha41575 commented:
/ok-to-test

natasha41575 commented:
> have you found an existing test that we can augment or duplicate to test this change?

@seh https://github.com/kubernetes-sigs/kustomize/blob/master/api/krusty/fnplugin_test.go has some tests for KRM exec functions that you may be able to model after.

aabouzaid commented:
@natasha41575 I've created the `DeepCopy`, and also added 2 tests one with just the base and another with overly (I've updated the PR body with the test results before and after the fix).

aabouzaid commented:
@KnVerey @natasha41575 could you please take a look at this PR? It's ready.

aabouzaid commented:
@KnVerey @natasha41575 could we at least run the CI pipeline to make sure everything as expected :sweat_smile:

KnVerey reviewed (commented):
Can you please also rebase when you address the comments? 
Re: the tests, Github isn't even offering me the button to run them for some reason. I've never seen that before, and I'll look into it if it doesn't start working after you push again.

natasha41575 commented:
@aabouzaid do you plan to continue this PR? It looks like we may need this functionality for https://github.com/kubernetes-sigs/kustomize/issues/3980, which @annasong20 is working on.

aabouzaid commented:
@natasha41575 Yes, I will try to finish it over the weekend.

aabouzaid commented on api/internal/plugins/loader/loader.go:
```diff
@@ -47,6 +47,18 @@ func (l *Loader) Config() *types.PluginConfig {
+		FnpLoadingOptions:  l.pc.FnpLoadingOptions,
+		HelmConfig:         l.pc.HelmConfig,
+	}
+	l.SetWorkDir(ldr.Root())
```
fixed :+1:

aabouzaid commented on api/internal/plugins/loader/loader.go:
```diff
@@ -47,6 +47,18 @@ func (l *Loader) Config() *types.PluginConfig {
 }
 
+// Copy the actual values of loader instead using the pointer values.
+func (l Loader) DeepCopy(ldr ifc.Loader) Loader {
```
Fixed :+1: 
I made it more specific so we don't need to copy unneeded/unused parts and in the same time make the naming clear.

aabouzaid commented:
@KnVerey @natasha41575 I believe that we are ready to go.
I've renamed the deep copy method to be more specific and added a test for it.

aabouzaid commented:
/assign @natasha41575

aabouzaid commented:
I just fixed the doc comment for the new methods.

annasong20 commented:
/assign

annasong20 commented on api/internal/plugins/loader/loader.go:
```diff
@@ -47,8 +47,18 @@ func (l *Loader) Config() *types.PluginConfig {
-// SetWorkDir sets the working directory for this loader's plugins
-func (l *Loader) SetWorkDir(wd string) {
+// DeepCopyPluginConfig makes a full copy the actual values of PluginConfig.
+func (l *Loader) DeepCopyPluginConfig() {
```
I'm behind this implementation because it fixes issue #4347. Ideally, though, we'd have a solution that's easier to understand. 

Need @KnVerey's feedback on this, but what about writing a 
```
func (l *Loader) CopyLoader() *Loader
``` 
method that only copies the relevant fields?

For instance, we 
* shouldn't make a deep copy of `fs filesys.FileSystem` because we want our `Loader`s to run on the same file system
* should copy `rf *resmap.Factory` because we [don't make a copy of it in the factory](https://github.com/kubernetes-sigs/kustomize/blob/master/api/internal/plugins/loader/loader.go#L40) and [expose it again here](https://github.com/kubernetes-sigs/kustomize/blob/master/api/internal/plugins/loader/loader.go#L213). We can do so by defining a `Copy()` method on `resmap.Factory` that we call.
* should definitely copy `pc *types.PluginConfig`, by defining `Copy()` methods at the `types.PluginConfig` and `FnpLoadingOptions` levels.

Another easier, though less safe, alternative is to store `struct`s instead of pointers to them. For example, in `Loader` we can store `pc types.PluginConfig` instead of `pc *types.PluginConfig` and thus we can use the value copy that go gives us. This wouldn't give us a complete copy, but would fix the krm issue.

aabouzaid replied:
That makes sense :+1: 

But if we agreed on the fact that it's not a deep copy ... and follow the same logic, isn't calling it `copy` or `new` would also be confusing? :thinking: 

Because the `Loader` struct fields have some pointers. Which is the root cause for the issue here (the nested cross-reference between copies).

And for the same reason (the `Loader` struct uses pointers), we cannot make it like the one in `fileloader` because we need to update a nested value in the struct.

So my suggestion as in the commit [proposal v2](https://github.com/kubernetes-sigs/kustomize/pull/4654/commits/05a752fba4c6b7c5590bb349e7c4c2947860cd8c) where it returns the loader after setting the working dir.

It's not `copy` and not `new` because, as mentioned, the cross-referencing.

Please correct me if I got it wrong.

annasong20 replied:
I think your current `LoaderWithWorkingDir` solution implements @KnVerey's 2nd suggestion spot on.

annasong20 commented on api/internal/plugins/loader/loader_test.go:
```diff
@@ -78,3 +78,17 @@ func TestLoader(t *testing.T) {
+	pLdrCopy := *pLdr
+	pLdrCopy.DeepCopyPluginConfig()
+	pLdrCopy.SetPluginConfigWorkingDir("/tmp/dummy")
+	if pLdrCopy.Config().FnpLoadingOptions.WorkingDir != "/tmp/dummy" {
```
I see you used `assert` in the integration test! If you use that here as well, these 3 lines can become 1. You can use `assert.Equal()` or `assert.Equalf()`. 

I also tend to use [the require package](https://pkg.go.dev/github.com/stretchr/testify/require), which is the same as `assert`, but ends the test on failure. Doesn't make a difference in your case, though.

annasong20 commented on api/internal/plugins/loader/loader_test.go:
```diff
@@ -78,3 +78,17 @@ func TestLoader(t *testing.T) {
 	}
 }
+
+func TestLoaderSetPluginConfigWorkingDir(t *testing.T) {
```
What's more is that we're testing the whether a deep enough copy was made. Can we change the test name to reflect that? For example, something like `TestDeepCopyPluginConfigWorkingDir`. 

Note that if we change the copy method name, we'll need to change the test name.

annasong20 replied:
Yes, apologies, I misread the test code!

annasong20 reviewed (commented):
@aabouzaid You should probably wait for @KnVerey's input before addressing the larger changes. 

I focused on your implementation this round. Will take a closer look at your tests in a later pass. Thanks for your work!

annasong20 commented:
/ok-to-test

annasong20 commented:
/retest

KnVerey commented on api/internal/plugins/loader/loader_test.go:
```diff
@@ -78,3 +79,16 @@ func TestLoader(t *testing.T) {
+	require.Equal(t,
+		"/tmp/dummy",
+		npLdr.Config().FnpLoadingOptions.WorkingDir,
+		"plugin working dir is not set correctly")
```
Please add a second assertion that the original loader's working dir did not change

KnVerey commented on api/internal/plugins/loader/loader.go:
```diff
@@ -42,16 +42,24 @@ func NewLoader(
 
+// LoaderWithWorkingDir returns loader after setting its working directory.
+// NOTE: This is not really a new loader since some of the Loader struct fields are pointers.
+func (l *Loader) LoaderWithWorkingDir(wd string) *Loader {
```
Yes, this looks great, thanks!

KnVerey reviewed (commented):
/approve

@annasong20 go ahead and lgtm once you're satisfied your comments are addressed

Thank you to both of you for your work on this fix and its review!

aabouzaid commented on api/krusty/fnplugin_test.go:
```diff
@@ -41,6 +41,12 @@ spec:
 EOF
 `
 
+const krmEchoDotSh = `#!/bin/bash
```
@annasong20 Pretty good point :ok_hand:

aabouzaid commented on api/krusty/fnplugin_test.go:
```diff
@@ -217,6 +223,124 @@ spec:
+`)
+
+	m := th.Run(prod, o)
+	assert.NoError(t, err)
```
@annasong20 that's a good catch :clap: :grinning:

aabouzaid commented on api/krusty/fnplugin_test.go:
```diff
@@ -126,7 +132,7 @@ spec:
 }
 
-func TestFnExecGeneratorWithOverlay(t *testing.T) {
+func TestFnExecGeneratorInOverlay(t *testing.T) {
```
Actually, I've renamed the functions to make it clear and avoid all of that confusion.

For example, for the transformer, we now have:
- **TestFnExecTransformerInBase**: The transformer is in the base, and there is no overlay.
- **TestFnExecTransformerInBaseWithOverlay**: The transformer is in the base, and there is an overlay, but it has no transformer.
- **TestFnExecTransformerInOverlay**: The transformer is in the overlay and not in the base.

And the generator follows the same pattern.

annasong20 replied:
Fantastic!

aabouzaid commented:
I've addressed all comments. So we are ready to go. :heavy_check_mark: 
I will rebase and update the commit messages once no more changes are required.

I see that there is alredy an error in the CI pipeline but it has nothing to do with this PR.
It's in `TestRemoteLoad_LocalProtocol` (it's related to this [line](https://github.com/kubernetes-sigs/kustomize/blob/7ee6dd551dc141deeb11a2efcc90e36df4f24ab5/api/krusty/remoteloader_test.go#L65) where I see the same error in other [commits pipeline runs](https://github.com/kubernetes-sigs/kustomize/actions/runs/3300702490/jobs/5445464576))

annasong20 commented on api/krusty/fnplugin_test.go:
```diff
@@ -41,7 +41,19 @@ spec:
 `
 
-func TestFnExecGenerator(t *testing.T) {
+const krmTransformerDotSh = `#!/bin/bash
```
Thank you for making this change!

annasong20 commented:
/lgtm

aabouzaid commented:
@annasong20 thanks for reviewing this PR :ok_hand: 
The last step was to rebase and squash the last 2 commits to have a meaningful message :sweat_smile:

annasong20 commented:
> @annasong20 thanks for reviewing this PR 👌 The last step was to rebase and squash the last 2 commits to have a meaningful message 😅

Totally on me. I didn't realize it would auto-merge... Thanks for bringing this up!
